File: telephony/telephonyothertest/telephonycpptest/sim/sim_reliability_test/src/sim_reliability_test.cpp
/*
 * Copyright (C) 2021 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <cstring>
#include <fstream>
#include "sim_test.h"
#include "sim_set_test.h"

using namespace testing::ext;
using namespace OHOS;
using namespace OHOS::Telephony;

const int32_t SimTest::LOOP_NUMBER = 1000;
const int32_t SimTest::SLOTID_0 = 0;
const int32_t SimTest::SLOTID_MINUS = -1;
const int32_t SimTest::OTHER_ABNORMAL = -2;
const int32_t SimTest::OPEN = 1;
const int32_t SimTest::CLOSE = 0;
const unsigned int SimTest::PHONE_NUMBER_LENGTH = 11;

/**
 * @tc.number   Telephony_Sim_SetVoiceMailInfo_0900
 * @tc.name     The test executed the SetVoiceMailInfo interface setting function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Function test
 */
HWTEST_F(SimSetVoiceMailTest, Telephony_Sim_SetVoiceMailInfo_0900, Function | MediumTest | Level1)
{
    std::u16string mailName = Str8ToStr16("123#@\%_ABCDE");
    std::u16string mailNumber = Str8ToStr16("01234567890123456789");
    for (int32_t loopNumber = 0; loopNumber < SimTest::LOOP_NUMBER; loopNumber++) {
        bool ret = g_proxy.GetHandler()->SetVoiceMailInfo(SimTest::SLOTID_0, mailName, mailNumber);
        ASSERT_TRUE(ret);
        std::u16string retMailNumber = g_proxy.GetHandler()->GetVoiceMailNumber(SimTest::SLOTID_0);
        ASSERT_STREQ(Str16ToStr8(retMailNumber).c_str(), Str16ToStr8(mailNumber).c_str());
        std::u16string retMailName = g_proxy.GetHandler()->GetVoiceMailIdentifier(SimTest::SLOTID_0);
        ASSERT_STREQ(Str16ToStr8(retMailName).c_str(), Str16ToStr8(mailName).c_str());
    }
}

/**
 * @tc.number   Telephony_Sim_IsSimActive_0500
 * @tc.name     The test executed the IsSimActive interface query function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimTest, Telephony_Sim_IsSimActive_0500, Reliability | MediumTest | Level3)
{
    for (int32_t loopNumber = 0; loopNumber < LOOP_NUMBER; loopNumber++) {
        bool result = g_proxy.GetHandler()->IsSimActive(SLOTID_0);
        ASSERT_TRUE(result);
    }
}

/**
 * @tc.number   Telephony_Sim_GetActiveSimAccountInfoList_0400
 * @tc.name     The test executed the IsSimActive interface query function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimSetTest, Telephony_Sim_GetActiveSimAccountInfoList_0400, Reliability | MediumTest | Level3)
{
    std::u16string cardName = Str8ToStr16("1234");
    std::u16string cardNumber = Str8ToStr16("1234");
    bool ret = g_proxy.GetHandler()->SetShowName(SimTest::SLOTID_0, cardName);
    ASSERT_TRUE(ret);
    ret = g_proxy.GetHandler()->SetShowNumber(SimTest::SLOTID_0, cardNumber);
    ASSERT_TRUE(ret);

    std::vector<OHOS::Telephony::IccAccountInfo> vecAccountInfo;
    bool findSlotIndex = false;
    std::u16string retIccId = g_proxy.GetHandler()->GetSimIccId(SimTest::SLOTID_0);
    for (int32_t loopNumber = 0; loopNumber < SimTest::LOOP_NUMBER; loopNumber++) {
        bool result = g_proxy.GetHandler()->GetActiveSimAccountInfoList(vecAccountInfo);
        ASSERT_TRUE(result);
        for (std::vector<OHOS::Telephony::IccAccountInfo>::iterator iter = vecAccountInfo.begin();
             iter != vecAccountInfo.end(); iter++) {
            if ((*iter).slotIndex == SimTest::SLOTID_0) {
                findSlotIndex = true;
                ASSERT_EQ((*iter).simId, SimTest::SLOTID_0);
                ASSERT_FALSE((*iter).isEsim);
                ASSERT_TRUE((*iter).isActive);
                ASSERT_STREQ(Str16ToStr8((*iter).iccId).c_str(), Str16ToStr8(retIccId).c_str());
                ASSERT_STREQ(Str16ToStr8((*iter).showName).c_str(), Str16ToStr8(cardName).c_str());
                ASSERT_STREQ(Str16ToStr8((*iter).showNumber).c_str(), Str16ToStr8(cardNumber).c_str());
            }
        }
        ASSERT_TRUE(findSlotIndex);
        findSlotIndex = false;
        usleep(100000);
    }
}

/**
 * @tc.number   Telephony_Sim_SetShowName_0500
 * @tc.name     The test executed the SetShowName interface query function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimSetTest, Telephony_Sim_SetShowName_0500, Reliability | MediumTest | Level3)
{
    std::u16string cardName = Str8ToStr16("1234");
    for (int32_t loopNumber = 0; loopNumber < SimTest::LOOP_NUMBER; loopNumber++) {
        bool ret = g_proxy.GetHandler()->SetShowName(SimTest::SLOTID_0, cardName);
        ASSERT_TRUE(ret);
    }
    std::u16string result = g_proxy.GetHandler()->GetShowName(SimTest::SLOTID_0);
    ASSERT_STREQ(Str16ToStr8(result).c_str(), Str16ToStr8(cardName).c_str());
}

/**
 * @tc.number   Telephony_Sim_GetShowName_0500
 * @tc.name     The test executed the GetShowName interface query function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimSetTest, Telephony_Sim_GetShowName_0500, Reliability | MediumTest | Level3)
{
    std::u16string cardName = Str8ToStr16("1234");
    bool ret = g_proxy.GetHandler()->SetShowName(SimTest::SLOTID_0, cardName);
    ASSERT_TRUE(ret);
    for (int32_t loopNumber = 0; loopNumber < SimTest::LOOP_NUMBER; loopNumber++) {
        std::u16string result = g_proxy.GetHandler()->GetShowName(SimTest::SLOTID_0);
        ASSERT_STREQ(Str16ToStr8(result).c_str(), Str16ToStr8(cardName).c_str());
    }
}

/**
 * @tc.number   Telephony_Sim_SetShowNumber_0500
 * @tc.name     The test executed the SetShowNumber interface query function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimTest, Telephony_Sim_SetShowNumber_0500, Reliability | MediumTest | Level3)
{
    std::u16string cardNumber = Str8ToStr16("1234");
    for (int32_t loopNumber = 0; loopNumber < LOOP_NUMBER; loopNumber++) {
        bool ret = g_proxy.GetHandler()->SetShowNumber(SLOTID_0, cardNumber);
        ASSERT_TRUE(ret);
        std::u16string result = g_proxy.GetHandler()->GetShowNumber(SLOTID_0);
        ASSERT_STREQ(Str16ToStr8(result).c_str(), Str16ToStr8(cardNumber).c_str());
    }
}

/**
 * @tc.number   Telephony_Sim_GetShowNumber_0400
 * @tc.name     The test executed the GetShowNumber interface query function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimTest, Telephony_Sim_GetShowNumber_0400, Reliability | MediumTest | Level3)
{
    std::u16string cardNumber = Str8ToStr16("1234");
    for (int32_t loopNumber = 0; loopNumber < LOOP_NUMBER; loopNumber++) {
        bool ret = g_proxy.GetHandler()->SetShowNumber(SLOTID_0, cardNumber);
        ASSERT_TRUE(ret);
        std::u16string result = g_proxy.GetHandler()->GetShowNumber(SLOTID_0);
        ASSERT_STREQ(Str16ToStr8(result).c_str(), Str16ToStr8(cardNumber).c_str());
    }
}

/**
 * @tc.number   Telephony_Sim_GetOperatorConfigs_0400
 * @tc.name     The test executed the GetOperatorConfigs interface query function 1000 times,
 *              expecting to execute successfully each time.
 *              Prerequisites: Mobile card.Manually push the XML configuration file corresponding to the
 *              mobile card to the 'data/OperatorConfig' path on the development board.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimTest, Telephony_Sim_GetOperatorConfigs_0400, Reliability | MediumTest | Level3)
{
    OHOS::Telephony::OperatorConfig poc;
    for (int32_t loopNumber = 0; loopNumber < LOOP_NUMBER; loopNumber++) {
        bool ret = g_proxy.GetHandler()->GetOperatorConfigs(SLOTID_0, poc);
        ASSERT_TRUE(ret);
        ASSERT_FALSE(poc.configValue.empty());
    }
}

/**
 * @tc.number   Telephony_Sim_GetSimTelephoneNumber_0400
 * @tc.name     The test executed the GetSimTelephoneNumber interface query function 1000 times,
 *              expecting to execute successfully each time.Prerequisite: The card number has been
 *              written into the SIM card.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimTest, Telephony_Sim_GetSimTelephoneNumber_0400, Reliability | MediumTest | Level3)
{
    for (int32_t loopNumber = 0; loopNumber < LOOP_NUMBER; loopNumber++) {
        std::u16string ret = g_proxy.GetHandler()->GetSimTelephoneNumber(SLOTID_0);
        ASSERT_STRNE(Str16ToStr8(ret).c_str(), "");
        // The length of a mobile phone number in mainland China is greater than or equal to 11 digits.
        ASSERT_GE(Str16ToStr8(ret).length(), PHONE_NUMBER_LENGTH);
    }
}

/**
 * @tc.number   Telephony_Sim_GetVoiceMailIdentifier_0400
 * @tc.name     The test executed the GetSimTelephoneNumber interface query function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimTest, Telephony_Sim_GetVoiceMailIdentifier_0400, Reliability | MediumTest | Level3)
{
    for (int32_t loopNumber = 0; loopNumber < LOOP_NUMBER; loopNumber++) {
        std::u16string ret = g_proxy.GetHandler()->GetVoiceMailIdentifier(SLOTID_0);
        ASSERT_STRNE(Str16ToStr8(ret).c_str(), "");
    }
}

/**
 * @tc.number   Telephony_Sim_GetVoiceMailNumber_0400
 * @tc.name     The test executed the GetVoiceMailNumber interface query function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimTest, Telephony_Sim_GetVoiceMailNumber_0400, Reliability | MediumTest | Level3)
{
    for (int32_t loopNumber = 0; loopNumber < LOOP_NUMBER; loopNumber++) {
        std::u16string ret = g_proxy.GetHandler()->GetVoiceMailNumber(SLOTID_0);
        ASSERT_STRNE(Str16ToStr8(ret).c_str(), "");
    }
}

/**
 * @tc.number   Telephony_Sim_UnlockPin2_0700
 * @tc.name     Loop to call UnlockPin2 interface for 1000 times and get the retValueurn value,
 *              and count the number of failures.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimTest, Telephony_Sim_UnlockPin2_0700, Reliability | MediumTest | Level3)
{
    OHOS::Telephony::LockStatusResponse response;
    std::string pin2 = "123";

    for (int32_t loopNumber = 0; loopNumber < LOOP_NUMBER; loopNumber++) {
        bool retValue = g_proxy.GetHandler()->UnlockPin2(SLOTID_0, Str8ToStr16(pin2), response);
        ASSERT_TRUE(retValue);
        ASSERT_EQ(response.result, OTHER_ABNORMAL);
    }
}

/**
 * @tc.number   Telephony_Sim_UnlockPuk2_0900
 * @tc.name     Loop to call UnlockPuk2 interface for 1000 times and get the retValueurn value,
 *              and count the number of failures.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimTest, Telephony_Sim_UnlockPuk2_0900, Reliability | MediumTest | Level3)
{
    OHOS::Telephony::LockStatusResponse response;
    std::u16string pin2 = Str8ToStr16("1234");
    std::u16string puk2 = Str8ToStr16("12345678");

    for (int32_t loopNumber = 0; loopNumber < LOOP_NUMBER; loopNumber++) {
        bool retValue = g_proxy.GetHandler()->UnlockPuk2(SLOTID_0, pin2, puk2, response);
        ASSERT_TRUE(retValue);
        ASSERT_EQ(response.result, OTHER_ABNORMAL);
    }
}

/**
 * @tc.number   Telephony_Sim_AlterPin2_0900
 * @tc.name     Loop to call UnlockPuk2 interface for 1000 times and get the retValueurn value,
 *              and count the number of failures.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimTest, Telephony_Sim_AlterPin2_0900, Reliability | MediumTest | Level3)
{
    OHOS::Telephony::LockStatusResponse response;
    std::u16string oldPin2 = Str8ToStr16("123");
    std::u16string newPin2 = Str8ToStr16("1234");

    for (int32_t loopNumber = 0; loopNumber < LOOP_NUMBER; loopNumber++) {
        bool retValue = g_proxy.GetHandler()->AlterPin2(SLOTID_MINUS, newPin2, oldPin2, response);
        ASSERT_FALSE(retValue);
    }
}

/**
 * @tc.number   Telephony_Sim_SetActiveSim_0600
 * @tc.name     The test executed the IsSimActive interface setting function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimSetTest, Telephony_Sim_SetActiveSim_0600, Reliability | MediumTest | Level3)
{
    for (int32_t loopNumber = 0; loopNumber < SimTest::LOOP_NUMBER / 2; loopNumber++) {
        bool ret = g_proxy.GetHandler()->SetActiveSim(SimTest::SLOTID_0, SimTest::CLOSE);
        ASSERT_TRUE(ret);
        bool result = g_proxy.GetHandler()->SetActiveSim(SimTest::SLOTID_0, SimTest::OPEN);
        ASSERT_TRUE(result);
    }
}

/**
 * @tc.number   Telephony_Sim_GetSimAccountInfo_0400
 * @tc.name     The test executed the GetSimAccountInfo interface query function 1000 times,
 *              expecting to execute successfully each time.
 * @tc.desc     Reliability test
 */
HWTEST_F(SimSetTest, Telephony_Sim_GetSimAccountInfo_0400, Reliability | MediumTest | Level3)
{
    OHOS::Telephony::IccAccountInfo simInfo;
    std::u16string cardName = Str8ToStr16("1234");
    std::u16string cardNumber = Str8ToStr16("1234");
    bool setNameResult = g_proxy.GetHandler()->SetShowName(SimTest::SLOTID_0, cardName);
    ASSERT_TRUE(setNameResult);
    bool setNumResult = g_proxy.GetHandler()->SetShowNumber(SimTest::SLOTID_0, cardNumber);
    ASSERT_TRUE(setNumResult);
    std::u16string retIccId = g_proxy.GetHandler()->GetSimIccId(SimTest::SLOTID_0);
    bool ret = g_proxy.GetHandler()->SetActiveSim(SimTest::SLOTID_0, SimTest::CLOSE);
    ASSERT_TRUE(ret);

    for (int32_t loopNumber = 0; loopNumber < SimTest::LOOP_NUMBER; loopNumber++) {
        bool result = g_proxy.GetHandler()->GetSimAccountInfo(SimTest::SLOTID_0, simInfo);
        ASSERT_TRUE(result);
        ASSERT_EQ(simInfo.simId, SimTest::SLOTID_0);
        ASSERT_EQ(simInfo.slotIndex, SimTest::SLOTID_0);
        ASSERT_FALSE(simInfo.isEsim);
        ASSERT_FALSE(simInfo.isActive);
        ASSERT_STREQ(Str16ToStr8(simInfo.iccId).c_str(), Str16ToStr8(retIccId).c_str());
        ASSERT_STREQ(Str16ToStr8(simInfo.showName).c_str(), Str16ToStr8(cardName).c_str());
        ASSERT_STREQ(Str16ToStr8(simInfo.showNumber).c_str(), Str16ToStr8(cardNumber).c_str());
        usleep(100000);
    }
}

File: telephony/handcrafttest/handcraftjstest/sim/sim_reliability_test/entry/src/main/js/test/SimReliability.test.js
/**
 * Copyright (C) 2021 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import sim from '@ohos.telephony.sim';
import * as env from './lib/Const';
import { describe, it, expect, Core } from 'deccjsunit/index';

describe('SimReliabilityTest', function () {

  // set timeout
  const core = Core.getInstance();
  const config = core.getDefaultService('config');
  config.timeout = 30 * 1000;

  /**
   *
   * @tc.number  Telephony_Sim_getOperatorConfigs_Async_04001
   * @tc.name    Call getOperatorConfigs, slotId is 0,10 times, and OperatorConfig is not expected to be empty
   * @tc.desc    Reliability test
   */
  it('Telephony_Sim_getOperatorConfigs_Async_04001', 0, async function (done) {
    const recursive = function (n) {
      if (n <= 0) {
        console.log('Telephony_Sim_getOperatorConfigs_Async_04001 finish');
        done();
        return;
      }
      console.log(`Telephony_Sim_getOperatorConfigs_Async_04001 run ${env.GENERAL_RUN_TIMES - n + 1} times`);
      sim.getOperatorConfigs(env.SLOTID0, (err, data) => {
        if (err) {
          console.log(`Telephony_Sim_getOperatorConfigs_Async_04001 fail, err : ${err.message}`);
          expect().assertFail();
          done();
          return;
        }
        if (data.length === 0) {
          expect(data.length !== 0).assertTrue();
          done();
        } else {
          recursive(n - 1);
        }
      });
    };
    recursive(env.GENERAL_RUN_TIMES);
  });

  /**
   * @tc.number  Telephony_Sim_getOperatorConfigs_Promise_04001
   * @tc.name    Call getOperatorConfigs, slotId is 0,10 times, and OperatorConfig is not expected to be empty
   * @tc.desc    Reliability test
   */
  it('Telephony_Sim_getOperatorConfigs_Promise_04001', 0, async function (done) {
    for (let index = 0; index < env.GENERAL_RUN_TIMES; index++) {
      console.log(`Telephony_Sim_getOperatorConfigs_Promise_04001 run ${index + 1} times`);
      try {
        let data = await sim.getOperatorConfigs(env.SLOTID0);
        expect(data.length !== 0).assertTrue();
      } catch (err) {
        console.log('Telephony_Sim_getOperatorConfigs_Promise_04001 fail');
        expect().assertFail();
        done();
        return;
      }
    }
    console.log('Telephony_Sim_getOperatorConfigs_Promise_04001 finish');
    done();
  });

  /**
   * @tc.number  Telephony_Sim_getSimTelephoneNumber_Async_04001
   * @tc.name    The getSimTelephoneNumber interface is called with the slotId parameter 0 and 10 times.
   *             The expected return value is not null
   * @tc.desc    Reliability test
   */
  it('Telephony_Sim_getSimTelephoneNumber_Async_04001', 0, async function (done) {
    const recursive = function (n) {
      if (n <= 0) {
        console.log('Telephony_Sim_getSimTelephoneNumber_Async_04001 finish');
        done();
        return;
      }
      console.log(`Telephony_Sim_getSimTelephoneNumber_Async_04001 run ${env.GENERAL_RUN_TIMES - n + 1} times`);
      sim.getSimTelephoneNumber(env.SLOTID0, (err, data) => {
        if (err) {
          console.log(`Telephony_Sim_getSimTelephoneNumber_Async_04001 fail, err : ${err.message}`);
          expect().assertFail();
          done();
          return;
        }
        expect(data.length !== 0).assertTrue();
        expect(data.length >= 11).assertTrue();
        recursive(n - 1);
      });
    };
    recursive(env.GENERAL_RUN_TIMES);
  });

  /**
   * @tc.number  Telephony_Sim_getSimTelephoneNumber_Promise_04001
   * @tc.name   The getSimTelephoneNumber interface is called with the slotId
   *            parameter 0 and 10 times. The expected return value is not null
   * @tc.desc    Reliability test
   */
  it('Telephony_Sim_getSimTelephoneNumber_Promise_04001', 0, async function (done) {
    for (let index = 0; index < env.GENERAL_RUN_TIMES; index++) {
      console.log(`Telephony_Sim_getSimTelephoneNumber_Promise_04001 run ${index + 1} times`);
      try {
        let data = await sim.getSimTelephoneNumber(env.SLOTID0);
        expect(data.length !== 0).assertTrue();
        expect(data.length >= 11).assertTrue();
      } catch (err) {
        console.log(`Telephony_Sim_getSimTelephoneNumber_Promise_04001 fail, err : ${err.message}`);
        expect().assertFail();
        done();
        return;
      }
    }
    console.log('Telephony_Sim_getSimTelephoneNumber_Promise_04001 finish');
    done();
  });

  /**
   * @tc.number  Telephony_Sim_getVoiceMailIdentifier_Async_04001
   * @tc.name    Test the stability of the getVoiceMailIdentifier (AsyncCallback) interface
   * @tc.desc    Reliability test
   */
  it('Telephony_Sim_getVoiceMailIdentifier_Async_04001', 0, async function (done) {
    const recursive = function (n) {
      if (n <= 0) {
        console.log('Telephony_Sim_getVoiceMailIdentifier_Async_04001 finish');
        done();
        return;
      }
      console.log(`Telephony_Sim_getVoiceMailIdentifier_Async_04001 run ${env.GENERAL_RUN_TIMES - n + 1} times`);
      sim.setVoiceMailInfo(env.SLOTID0, env.MAIL_NAME_LEN12, env.MAIL_NUMBER_LEN20, err => {
        if (err) {
          console.log(`Telephony_Sim_getVoiceMailIdentifier_Async_04001 setVoiceMailInfo err: ${err.message}`);
          expect().assertFail();
          done();
          return;
        }
        sim.getVoiceMailIdentifier(env.SLOTID0, (err, data) => {
          if (err || data !== env.MAIL_NAME_LEN12) {
            console.log(
              `Telephony_Sim_getVoiceMailIdentifier_Async_04001 test err : ${err && err.message}, identifier: ${data}`
            );
            expect().assertFail();
            done();
            return;
          }
          recursive(n - 1);
        });
      });
    };
    recursive(env.GENERAL_RUN_TIMES);
  });

  /**
   * @tc.number Telephony_Sim_getVoiceMailIdentifier_Promise_04001
   * @tc.name   Call the getVoiceMailIdentifier interface with slotId input parameter 0, 10 times
   *            expecting that if getSimOperatorNumeric (0) is equal to 46000 or 46002 or 46004 or 46007,
   *            the return value is 12599 Otherwise, the return value is not null
   * @tc.desc   Reliability test
   */
  it('Telephony_Sim_getVoiceMailIdentifier_Promise_04001', 0, async function (done) {
    try {
      for (let i = 0; i < env.GENERAL_RUN_TIMES; i++) {
        console.log(`Telephony_Sim_getVoiceMailIdentifier_Promise_04001 run ${i + 1} times`);
        await sim.setVoiceMailInfo(env.SLOTID0, env.MAIL_NAME_LEN12, env.MAIL_NUMBER_LEN20);
        const indentifier = await sim.getVoiceMailIdentifier(env.SLOTID0);
        if (indentifier !== env.MAIL_NAME_LEN12) {
          console.log(`Telephony_Sim_getVoiceMailIdentifier_Promise_04001 test fail, indentifier: ${indentifier}`);
          expect(indentifier === env.MAIL_NAME_LEN12).assertTrue();
          done();
          return;
        }
      }
      console.log('Telephony_Sim_getVoiceMailIdentifier_Promise_04001 finish');
    } catch (error) {
      console.log(`Telephony_Sim_getVoiceMailIdentifier_Promise_04001 error: ${error.message}`);
      expect().assertFail();
    }
    done();
  });

  /**
   * @tc.number  Telephony_Sim_getVoiceMailNumber_Async_04001
   * @tc.name    The getVoiceMailNumber interface is called with the slotId input parameter 0, 10 times,
   *             expecting that if getSimOperatorNumeric (0) is equal to 46000 or 46002 or
   *             46004 or 46007, the return value is 12599
   *             Otherwise, the return value is not null
   * @tc.desc    Reliability test
   */
  it('Telephony_Sim_getVoiceMailNumber_Async_04001', 0, async function (done) {
    const recursive = function (n) {
      if (n <= 0) {
        console.log('Telephony_Sim_getVoiceMailNumber_Async_04001 finish');
        done();
        return;
      }
      console.log(`Telephony_Sim_getVoiceMailNumber_Async_04001 run ${env.GENERAL_RUN_TIMES - n + 1} times`);
      sim.setVoiceMailInfo(env.SLOTID0, env.MAIL_NAME_LEN12, env.MAIL_NUMBER_LEN20, err => {
        if (err) {
          console.log(`Telephony_Sim_getVoiceMailNumber_Async_04001 setVoiceMailInfo err: ${err.message}`);
          expect().assertFail();
          done();
          return;
        }
        sim.getVoiceMailNumber(env.SLOTID0, (err, data) => {
          if (err || data !== env.MAIL_NUMBER_LEN20) {
            console.log(
              `Telephony_Sim_getVoiceMailNumber_Async_04001 test err : ${err && err.message}, identifier: ${data}`
            );
            expect().assertFail();
            done();
            return;
          }
          recursive(n - 1);
        });
      });
    };
    recursive(env.GENERAL_RUN_TIMES);
  });

  /**
   * @tc.number  Telephony_Sim_getVoiceMailNumber_Promise_04001
   * @tc.name    Call the getVoiceMailNumber interface with slotId input parameter 0, 10 times
   *             expecting that if getSimOperatorNumeric (0) is equal to 46000 or 46002 or 46004 or 46007,
   *             the return value is 12599 Otherwise, the return value is not null
   * @tc.desc    Reliability test
   */
  it('Telephony_Sim_getVoiceMailNumber_Promise_04001', 0, async function (done) {
    try {
      for (let i = 0; i < env.GENERAL_RUN_TIMES; i++) {
        console.log(`Telephony_Sim_getVoiceMailNumber_Promise_04001 run ${i + 1} times`);
        await sim.setVoiceMailInfo(env.SLOTID0, env.MAIL_NAME_LEN12, env.MAIL_NUMBER_LEN20);
        const number = await sim.getVoiceMailNumber(env.SLOTID0);
        if (number !== env.MAIL_NUMBER_LEN20) {
          console.log(`Telephony_Sim_getVoiceMailNumber_Promise_04001 test fail, number: ${number}`);
          expect(number === env.MAIL_NUMBER_LEN20).assertTrue();
          done();
          return;
        }
      }
      console.log('Telephony_Sim_getVoiceMailNumber_Promise_04001 finish');
    } catch (error) {
      console.log(`Telephony_Sim_getVoiceMailNumber_Promise_04001 error: ${error.message}`);
      expect().assertFail();
    }
    done();
  });

  /**
   * @tc.number   Telephony_Sim_setVoiceMailInfo_Async_09001
   * @tc.name     Test the reliability of the setVoiceMailInfo async callback interface
   * @tc.desc     Reliability test
   */
  it('Telephony_Sim_setVoiceMailInfo_Async_09001', 0, function (done) {
    const recursive = function (n) {
      if (n <= 0) {
        console.log('Telephony_Sim_setVoiceMailInfo_Async_09001 test finish.');
        done();
        return;
      }
      console.log(`Telephony_Sim_setVoiceMailInfo_Async_09001 run ${env.GENERAL_RUN_TIMES - n + 1} times`);
      sim.setVoiceMailInfo(env.SLOTID0, env.MAIL_NAME_LEN12, env.MAIL_NUMBER_LEN20, error => {
        if (error) {
          console.log(`Telephony_Sim_setVoiceMailInfo_Async_09001 setVoiceMailInfo error: ${error.message}`);
          expect().assertFail();
          done();
          return;
        }
        sim.getVoiceMailNumber(env.SLOTID0, (error, value) => {
          if (error) {
            console.log(`Telephony_Sim_setVoiceMailInfo_Async_09001 getVoiceMailNumber error: ${error.message}`);
            expect().assertFail();
            done();
            return;
          }
          if (value !== env.MAIL_NUMBER_LEN20) {
            console.log(`Telephony_Sim_setVoiceMailInfo_Async_09001 getVoiceMailNumber value: ${value}`);
            expect().assertFail();
            done();
            return;
          }
          sim.getVoiceMailIdentifier(env.SLOTID0, (error, value) => {
            if (error || value !== env.MAIL_NAME_LEN12) {
              console.log(`Telephony_Sim_setVoiceMailInfo_Async_09001 error: ${error.message}, identifier: ${value}`);
              expect().assertFail();
              done();
              return;
            }
            recursive(n - 1);
          });
        });
      });
    };
    recursive(env.GENERAL_RUN_TIMES);
  });

  /**
   * @tc.number   Telephony_Sim_setVoiceMailInfo_Promise_09001
   * @tc.name     Test the reliability of the setVoiceMailInfo promise interface
   * @tc.desc     Reliability test
   */
  it('Telephony_Sim_setVoiceMailInfo_Promise_09001', 0, async function (done) {
    try {
      for (let i = 0; i < env.GENERAL_RUN_TIMES; i++) {
        console.log(`Telephony_Sim_setVoiceMailInfo_Promise_09001 run ${i + 1} times`);
        await sim.setVoiceMailInfo(env.SLOTID0, env.MAIL_NAME_LEN12, env.MAIL_NUMBER_LEN20);
        const mailNumber = await sim.getVoiceMailNumber(env.SLOTID0);
        if (mailNumber !== env.MAIL_NUMBER_LEN20) {
          console.log(`Telephony_Sim_setVoiceMailInfo_Promise_09001 getVoiceMailNumber value: ${mailNumber}`);
          expect().assertFail();
          done();
          return;
        }
        const mailIdentifier = await sim.getVoiceMailIdentifier(env.SLOTID0);
        if (mailIdentifier !== env.MAIL_NAME_LEN12) {
          console.log(`Telephony_Sim_setVoiceMailInfo_Promise_09001 getVoiceMailIdentifier value: ${mailIdentifier}`);
          expect(mailIdentifier === env.MAIL_NAME_LEN12).assertTrue();
          done();
          return;
        }
      }
      console.log('Telephony_Sim_setVoiceMailInfo_Promise_09001 test finish');
    } catch (error) {
      console.log(`Telephony_Sim_setVoiceMailInfo_Promise_09001 error: ${error.message}`);
      expect().assertFail();
    }
    done();
  });
});